### ISSUE TITLE
Native ability to sort a <table>

### ISSUE THREAD

**Issue by scottkellum** [openui/open-ui#800]
Sorting a table is common functionality added to a table via JavaScript. It would be helpful to have this functionality natively added to tables.

### Related material:

- https://www.w3.org/WAI/ARIA/apg/patterns/table/examples/sortable-table/
- https://blog.logrocket.com/creating-react-sortable-table/
- https://stackoverflow.com/questions/14267781/sorting-html-table-with-javascript
- https://javascript.info/task/sortable-table
- https://webdesign.tutsplus.com/how-to-create-a-sortable-html-table-with-javascript--cms-92993t
- https://github.hubspot.com/sortable/docs/welcome/

Possible solution: add an attribute to a <th> of `sortable="true"` with a `::marker` of up/down (like <select>) for un-sorted, `▼` for descending, and `▲` for ascending.

**Comment by LeaVerou** [comment]
First, huge +1 on addressing the user need. Sorting, filtering etc *should* be possible declaratively, out of the box. This would also allow extensions to enable this functionality on *any* table on the web.

Nit: `sortable="true"` is not how HTML does booleans. If it's boolean, it should be `sortable` with no value (I also opened https://github.com/w3ctag/design-principles/issues/451 about this).

I imagine the first thing to be ironed out would be *how* do we sort? Lexicographically is rarely useful for strings, but do we really want to duck type numbers, with all the localization implications this ensues? Perhaps we also need an enumerated attribute on `<col>` about the type of sort? And whether ascending or descending is the default?

Regardless of which heuristic we pick as the default, we'd need to provide some kind of escape hatch for when it fails. For example, we could introduce a `value` attribute for `<td>`s, which would override whatever value is in `.textContent`.

**Comment by bkardell** [comment]
Yeah, totally agree that this is useful. If someone wanted to start research on it there are volumes for sure of different attempts at this in markup - it would be useful to look at what variations there are and why.

"HTML 5" early on had a `<datagrid>` element, it's hard to point to something that isn't misleading/misrepresentative there though because it was all in such flux, no single date might really "represent" it entirely.  Still https://web.archive.org/web/20050602074634/http://www.whatwg.org/specs/web-apps/current-work/#data-grids is one grab that seems "pretty good" from my recollections.  We can also look at the mailing lists and irc logs from then for developments and past discussions on that.

But also, these have been in other markups at least as long that tried to do "richer tables" of all kinds that I think are worth study and efforts... Even currently there are 70 custom element "tables" that show up in the recent HTTP Archive crawl, for example https://rainy-periwinkle.glitch.me/permalink/5c27a16fa9188e838ce164e76ebcb0045e555315c5e878b736c9631e776d526c.html

**Comment by dbaron** [comment]
Another set of documentation for existing table sorting features is https://meta.wikimedia.org/wiki/Help:Sorting which documents sortable tables in mediawiki (which are quite complicated).

**Comment by gregwhitworth** [comment]
@scottkellum on all table issues the Open UI CG agreed that an explainer in this CG makes sense. Would you be open to championing such an explainer for proposed solutions for the issues?

**Comment by scottkellum** [comment]
@gregwhitworth thanks for the update. I really wish I had time but both work and personal responsibilities have been absolutely crushing lately.

Apologies I feel like I dropped a big one on OpenUI's plate. I will likely have more time when my daughter starts daycare in 3 months if this can wait until then.

**Comment by keithamus** [comment]
In #903 I suggested:

> For sorting I propose having `<th sort=asc>`/`<th sort=desc>` to sort lexicographically and `<th sort=asc-numeric>`/`<td sort=desc-numeric>` to sort numerically. In addition perhaps having the IDL allow assigning a comparator function (`sort = (a, b) => {...}`) could be useful for some customisation of the sorting heuristic. Sorting could use slottable, where the table moves into a state that all rows/columns become slots that can re-organise based on the sort order.
> 
> I agree invokers could play a useful role here.

**Comment by LeaVerou** [comment]
> Invokers could play a useful role here for allowing userland to add their own controls to do the sorting, rather than injecting UI elements that then would need styling via something like `:marker`.

In the spirit of making common things easy, and complex things possible, I think the ideal would be if the UA created controls but the author could provide their own controls as well (via slots?). If all that's needed is to style the UA provided controls, `::marker` seems fine.

**Comment by jakearchibald** [comment]
I wonder how much crossover there is between "I want advanced table features" and "I'm not displaying all the data at once".

**Comment by Delapouite** [comment]
> I wonder how much crossover there is between "I want advanced table features" and "I'm not displaying all the data at once".

Two related WICG repos around this topic : 
- https://github.com/WICG/virtual-scroller about an hypothetic `<virtual-scroller>` standard element, archived in 2021 and pointing to this second repo : 
- https://github.com/WICG/display-locking which led to the adoption of `content-visibility`, `contain-intrinsic-size`… archived in 2022

**Comment by annevk** [comment]
HTML had sortable tables at one point: https://github.com/whatwg/html/commit/6db0d8d4e3456140de958c963afe9bb9ec7b6a25. And then it got removed due to lack of interest: https://github.com/whatwg/html/commit/59b7e2466c2b7c5c408a4963b05b13fd808aa07a. This is not the `datagrid` element, but a newer iteration of that. At this point that's over a decade old so maybe it's worth another shot, but I'd expect to see some kind of comparison to this in any proposal, including why it's going to succeed this time around.

**Comment by bkardell** [comment]
I guess this is a fairly decent grab from wayback  https://web.archive.org/web/20140531031702/http://www.whatwg.org/specs/web-apps/current-work/multipage/tabular-data.html#table-sorting-model - if we were to pick it up again
